src/lem/circuit.rs

```diff
@@ -56,9 +58,157 @@ use super::{
     Block, Ctrl, Func, Op, Tag, Var,
 };
 
+#[derive(Clone)]
+pub struct MultiFrame<'a, F: LurkField> {
+    pub func: &'a Func,
+    pub store: Option<&'a Store<F>>,
+    pub input: Option<Vec<Ptr<F>>>,
+    pub output: Option<Vec<Ptr<F>>>,
+    pub frames: Option<Vec<Frame<F>>>,
+    pub cached_witness: Option<WitnessCS<F>>,
+    pub count: usize,
+}
+
+impl<'a, F: LurkField> MultiFrame<'a, F> {
+    pub fn blank(func: &'a Func, count: usize) -> Self {
+        Self {
+            func,
+            store: None,
+            input: None,
+            output: None,
+            frames: None,
+            cached_witness: None,
+            count,
+        }
+    }
+
+    pub fn get_store(&self) -> &Store<F> {
+        self.store.expect("store missing")
+    }
+
+    pub fn from_frames(
+        func: &'a Func,
+        count: usize,
+        frames: &[Frame<F>],
+        store: &'a Store<F>,
+    ) -> Vec<Self> {
+        // `count` is the number of `Frames` to include per `MultiFrame`.
+        let total_frames = frames.len();
+        let n = total_frames / count + usize::from(total_frames % count != 0);
+        let mut multi_frames = Vec::with_capacity(n);
+
+        for chunk in frames.chunks(count) {
+            let mut inner_frames = Vec::with_capacity(count);
+
+            for frame in chunk {
+                inner_frames.push(frame.clone());
+            }
+
+            let last_frame = chunk.last().expect("chunk must not be empty");
+            let last_circuit_frame = inner_frames
+                .last()
+                .expect("chunk must not be empty")
+                .clone();
+
+            // Fill out the MultiFrame, if needed, and capture output of the final actual frame.
+            for _ in chunk.len()..count {
+                inner_frames.push(last_circuit_frame.clone());
+            }
+
+            let output = last_frame.output.clone();
+            let input = chunk[0].input.clone();
+            debug_assert!(!inner_frames.is_empty());
+
+            let mf = MultiFrame {
+                func,
+                store: Some(store),
+                input: Some(input),
+                output: Some(output),
+                frames: Some(inner_frames),
+                cached_witness: None,
+                count,
+            };
+
+            multi_frames.push(mf);
+        }
+
+        multi_frames
+    }
+
+    /// Make a dummy `MultiFrame`, duplicating `self`'s final `CircuitFrame`.
```

Comment:
This comment seems wrong. The function doesn't have a `self` parameter. I think you mean duplicating `circuit_frame`.